keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+  placement, the actual orchestration of optimized GPU-to-GPU communication often relies on topology awareness
+  within the machine learning framework itself (e.g., Ray, PyTorch Distributed. etc).  Kubernetes provides the
+  necessary foundation for these frameworks to operate efficiently, but the frameworks are ultimately responsible
+  for leveraging that foundation.
+* CNI plugins can leverage topology information to establish more efficient network paths,
+  reducing latency and increasing throughput. For instance, a CNI could prioritize connections
+  within the same availability zone or rack.
+* Distributed stateful applications, such as sharded databases, can use topology information to improve fault tolerance.
+  By spreading replicas and traffic across different failure domains (e.g., zones, racks), these applications can achieve
+  higher availability and resilience to failures in any one topology.
+
+Today, applications obtain topology information by using an init container to query the Kubernetes API for the underlying node,
+extracting the necessary data from the Node resource. As more workloads and use cases benefit from topology awareness,
+we aim to simplify access to this information for Pods via the Downward API.
+
+### Goals
+
+* Values from Node labels `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are made
+  available via downward API
+* Additional node labels can be made available via downward API using admission webhooks that mutate `pods/binding`.
+
+### Non-Goals
+
+* Exposing non-standard node labels by default
+* Enhnacements to topology-aware scheduling
+* Changes to standard topology labels in Kubernetes
+* Changes to downward API
+
+## Proposal
+
+Retreival of topology information will be achieved through the existing mechanism of passing down Pod labels using Downward API.
+A new admission plugin, `PodTopologyLabels`, will be added to kube-apiserver. When enabled, it will mutate the `pods/binding` subresource,
+adding topology labels that match those of the target Node. The Binding REST implementation will be updated to copy these labels
+from the `pods/binding` subresource to the assigned Pod's labels.
+
+Using the downward API to retrieve topology information will look similar to the following:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: downwardapi-example-
+spec:
+  containers:
+    - name: container
+      image: registry.k8s.io/busybox
+      command: ["sh", "-c"]
+      args:
+      - while true; do
+          if [[ -e /etc/podinfo/zone ]]; then
+            echo -en '\n\n'; cat /etc/podinfo/zone; fi;
+          sleep 5;
+        done;
+      volumeMounts:
+        - name: podinfo
+          mountPath: /etc/podinfo
+  volumes:
+    - name: podinfo
+      downwardAPI:
+        items:
+          - path: "zone"
+            fieldRef:
+              fieldPath: metadata.labels['topology.kubernetes.io/zone']
```

Comment:
what is the expected behavior when the label ended up missing?

Reply:
yes, I just want to make sure that the existing behavior on missing labels is OK for this KEP. PRR reviewers may want to know the effect of disabling the feature gate

Reply:
I believe expected behavior is that it just has an empty value in the Pod, it doesn't prevent Pods from running

Reply:
maybe mention it in PRR when saying what is the outcome for the end user when feature is disabled

Reply:
good suggestion, thanks Sergey

Reply:
The topology labels would still be available on already running Pods even if the feature is disabled though, updated PRR to make this clearer